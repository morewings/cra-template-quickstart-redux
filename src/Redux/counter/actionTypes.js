export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
