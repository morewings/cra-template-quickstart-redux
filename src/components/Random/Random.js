import React from 'react';
import {
  useGetRandomNumberQuery,
  useRandomNumber,
  useLoadingState,
} from 'features/random';
import classes from './Random.module.css';

const Random = () => {
  /** Loading state of random.org request from Redux store */
  const {isLoading, hasError, isFulfilled} = useLoadingState();

  /** Random number value */
  const number = useRandomNumber();

  /** Create incrementCounter action, using custom hook from feature */
  const getNumber = useGetRandomNumberQuery();

  /** Define pristine state condition, when user didn't do any actions */
  const isPristine = !isLoading && !hasError && !isFulfilled;

  return (
    <div className={classes.counter}>
      <h2 className={classes.header}>Async Random</h2>
      <button
        disabled={isLoading}
        className={classes.button}
        type="button"
        onClick={getNumber}>
        Get random number
      </button>
      {isPristine && <div>Click the button to get random number</div>}
      {isLoading && <div>Getting number</div>}
      {isFulfilled && (
        <div>
          Number from random.org: <strong>{number}</strong>
        </div>
      )}
      {hasError && <div>Ups...</div>}
    </div>
  );
};

export default Random;
