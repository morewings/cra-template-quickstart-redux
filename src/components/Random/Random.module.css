.counter {
  border: 1px solid lightgray;
  margin: 36px 24px;
  padding: 24px;
  text-align: center;
  width: 240px;
}

.header {
  font-size: 24px;
  font-weight: normal;
  margin: 0 0 12px;
}

.button {
  background: lightseagreen;
  border: none;
  border-radius: 5px;
  color: white;
  cursor: pointer;
  display: block;
  font-size: 16px;
  margin: 0 auto 24px;
  padding: 12px 24px;
  text-shadow: 1px 1px 1px rgba(0, 0, 0, 0.5);
}

.button:active {
  left: 1px;
  position: relative;
  top: 1px;
}
