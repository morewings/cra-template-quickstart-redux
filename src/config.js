export default Object.freeze({
  randomAPI:
    'https://www.random.org/integers/?num=1&min=1&max=99&col=1&base=10&format=plain&rnd=new',
});
