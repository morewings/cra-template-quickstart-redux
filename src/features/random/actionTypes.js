export const GET_RANDOM_NUMBER = 'GET_RANDOM_NUMBER';
