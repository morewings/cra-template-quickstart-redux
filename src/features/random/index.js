export {default as RandomReducer} from './RandomReducer';
export {default as useRandomAPI} from './selectors';
export {default as useActions} from './actionCreators';
