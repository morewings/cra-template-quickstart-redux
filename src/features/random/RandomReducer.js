import {GET_RANDOM_NUMBER} from './actionTypes';

const initialState = {
  number: undefined,
  isLoading: false,
  hasError: false,
  isFulfilled: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case `${GET_RANDOM_NUMBER}_PENDING`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: true,
        hasError: false,
        number: undefined,
      };

    case `${GET_RANDOM_NUMBER}_FULFILLED`:
      return {
        isFulfilled: true,
        isLoading: false,
        hasError: false,
        number: action.payload.data,
      };

    case `${GET_RANDOM_NUMBER}_REJECTED`:
      return {
        isFulfilled: false,
        isLoading: false,
        hasError: true,
        number: undefined,
      };

    default:
      return state;
  }
};
