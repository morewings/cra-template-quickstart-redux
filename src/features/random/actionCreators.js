import {useCallback} from 'react';
import {useDispatch} from 'react-redux';
import axios from 'axios';
import config from '../../config';
import {GET_RANDOM_NUMBER} from './actionTypes';

const useActions = () => {
  const dispatch = useDispatch();
  const getNumber = useCallback(
    () =>
      dispatch({
        type: GET_RANDOM_NUMBER,
        payload: axios.get(config.randomAPI),
      }),
    [dispatch]
  );
  return {getNumber};
};

export default useActions;
