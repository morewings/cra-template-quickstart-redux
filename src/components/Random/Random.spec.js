import React from 'react';
import {Provider} from 'react-redux';
import {render, fireEvent, waitFor} from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import promise from 'redux-promise-middleware';
import configureStore from 'redux-mock-store';
import {GET_RANDOM_NUMBER} from '../../features/random/actionTypes';
import Random from './Random';
import config from '../../config';

describe('components > Random', () => {
  /**
   * Create mock store
   * @see https://github.com/dmitry-zaets/redux-mock-store
   */
  const mockStore = configureStore([promise]);

  /**
   * Initialize axios mock adapter to mock API responses
   * @see https://github.com/ctimmerm/axios-mock-adapter
   */
  const mockAxios = new MockAdapter(axios);

  beforeEach(() => {
    mockAxios.resetHandlers();
  });

  /**
   * Provide table of values to run tests with
   * @see https://jestjs.io/docs/en/api#describeeachtablename-fn-timeout
   */
  describe.each`
    isLoading | hasError | isFulfilled
    ${false}  | ${false} | ${false}
    ${true}   | ${false} | ${false}
    ${false}  | ${true}  | ${false}
    ${false}  | ${false} | ${true}
  `('renders different store states', ({isLoading, hasError, isFulfilled}) => {
    it(`it renders when isLoading === ${isLoading} && hasError === ${hasError} && isFulfilled === ${isFulfilled}`, () => {
      const store = mockStore({
        random: {
          isLoading,
          hasError,
          isFulfilled,
          number: isFulfilled ? 1 : undefined,
        },
      });

      /**
       * `asFragment`:
       * @see https://testing-library.com/docs/react-testing-library/api#asfragment
       * `wrapper`
       * @see https://testing-library.com/docs/react-testing-library/api#wrapper
       */
      const {asFragment} = render(<Random />, {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      });

      /**
       * Basic snapshot test to make sure, that rendered component
       * matches expected footprint.
       */
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('dispatches an action sequence on button click', async () => {
    const store = mockStore({
      random: {
        isLoading: false,
        hasError: false,
        isFulfilled: false,
      },
    });

    /** Mock response from API */
    const response = 6;
    mockAxios.onGet(config.randomAPI).reply(200, response);

    /**
     * `getByRole`:
     * @see https://testing-library.com/docs/dom-testing-library/api-queries#byrole
     */
    const {getByRole} = render(<Random />, {
      wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
    });

    /**
     * Search for the button and make testing library click on it
     * @see https://testing-library.com/docs/react-testing-library/cheatsheet#events
     */
    fireEvent.click(getByRole('button'));

    /** First dispatched action should have _PENDING suffix */
    expect(store.getActions()[0]).toEqual({
      type: `${GET_RANDOM_NUMBER}_PENDING`,
    });

    await waitFor(() => {
      /** Second dispatched action should have _FULFILLED suffix */
      expect(store.getActions()[1].type).toEqual(
        `${GET_RANDOM_NUMBER}_FULFILLED`
      );

      /** Second dispatched action should deliver response from API */
      expect(store.getActions()[1].payload.data).toEqual(response);
    });
  });
});
