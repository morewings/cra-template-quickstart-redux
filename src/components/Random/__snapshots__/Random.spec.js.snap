// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Random renders different store states it renders when isLoading === false && hasError === false && isFulfilled === false 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <h2
      class="header"
    >
      Async Random
    </h2>
    <button
      class="button"
      type="button"
    >
      Get random number
    </button>
    <div>
      Click the button to get random number
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random renders different store states it renders when isLoading === false && hasError === false && isFulfilled === true 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <h2
      class="header"
    >
      Async Random
    </h2>
    <button
      class="button"
      type="button"
    >
      Get random number
    </button>
    <div>
      Number from random.org: 
      <strong>
        1
      </strong>
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random renders different store states it renders when isLoading === false && hasError === true && isFulfilled === false 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <h2
      class="header"
    >
      Async Random
    </h2>
    <button
      class="button"
      type="button"
    >
      Get random number
    </button>
    <div>
      Ups...
    </div>
  </div>
</DocumentFragment>
`;

exports[`components > Random renders different store states it renders when isLoading === true && hasError === false && isFulfilled === false 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <h2
      class="header"
    >
      Async Random
    </h2>
    <button
      class="button"
      disabled=""
      type="button"
    >
      Get random number
    </button>
    <div>
      Getting number
    </div>
  </div>
</DocumentFragment>
`;
